### PR TITLE
Knockoff example

### PR DESCRIPTION
**Pull request by alexblnn**
Adding knockoff example

### CODE REVIEW

File: examples/example_knockoffs.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+"""
+Controlled variable selection: why and how?
```
Comment (bthirion) [inline examples/example_knockoffs.py:2]:
```suggestion
Controlled variable selection with knokoffs
```

File: examples/example_knockoffs.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+# %%
+feature_names = [str(name) for name in data.feature_names]
+feature_names[:10]
+
+# %%
+data.target_names
+
+# %% [markdown]
+# ## Step 1: Selecting variables with the Lasso
+# We want to select variables that are relevant to the outcome, i.e. tumor charateristics that are associated with tumor malignance. We start off by applying a classical method using Lasso logistic regression and retaining variables with non-null coefficients:
```
Comment (bthirion) [inline examples/example_knockoffs.py:43]:
```suggestion
# We want to select variables that are relevant to the outcome, i.e. tumor charateristics that are associated with tumor malignance. We start off by applying a classical method using L1-regularized logistic regression and retaining variables with non-null coefficients:
```

File: examples/example_knockoffs.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+clf = LogisticRegressionCV(penalty="l1", solver="liblinear", max_iter=int(1e4))
+clf.fit(X, y)
+
+selected_lasso = np.where(np.abs(clf.coef_[0]) > 1e-6)[0]
+print(f" {len(selected_lasso)} features are selected by the Lasso:")
+print(np.array(feature_names)[np.abs(clf.coef_[0]) > 1e-6])
+
+# %% [markdown]
+# ## Step 2: Evaluating the rejection set
+# Since we do not have the ground truth, we cannot evaluate this selection set directly. To investigate the reliability of this method, we artificially increase the number of variables by adding synthetic noise features. These are completely irrelevant to the problem at hand.
```
Comment (bthirion) [inline examples/example_knockoffs.py:57]:
```suggestion
# Since we have no ground truth, we cannot evaluate this selection set directly. To investigate the reliability of this method, we artificially increase the number of variables by adding synthetic noise features. These are completely irrelevant to the prediction problem.
```

File: examples/example_knockoffs.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+# %%
+repeats_noise = 3  # Number of synthetic noisy sets to add
+
+X_to_shuff = X.copy()
+noises = [X]
```
Comment (bthirion) [inline examples/example_knockoffs.py:63]:
```suggestion
noise = [X]
```

File: examples/example_knockoffs.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+print(np.array(feature_names)[np.abs(clf.coef_[0]) > 1e-6])
+
+# %% [markdown]
+# ## Step 2: Evaluating the rejection set
+# Since we do not have the ground truth, we cannot evaluate this selection set directly. To investigate the reliability of this method, we artificially increase the number of variables by adding synthetic noise features. These are completely irrelevant to the problem at hand.
+
+# %%
+repeats_noise = 3  # Number of synthetic noisy sets to add
+
+X_to_shuff = X.copy()
```
Comment (bthirion) [inline examples/example_knockoffs.py:62]:
```suggestion
X_shuffle = X.copy()
```

File: examples/example_knockoffs.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+noises = [X]
+
+for k in range(repeats_noise):
+    np.random.shuffle(X_to_shuff.T)
```
Comment (bthirion) [inline examples/example_knockoffs.py:66]: You should add it to the noise ? Not done atm.

File: examples/example_knockoffs.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+print("Selected features with Knockoffs:")
+print(np.array(feature_names)[selected])
+
+# Count how many selected features are actually noise
+num_false_discoveries = np.sum(selected >= p)
+print(f"Knockoffs make at least {num_false_discoveries} False Discoveries")
+
+# %% [markdown]
+# Knockoffs select the same variable as previously and select **no fake variables**.
```
Comment (bthirion) [inline examples/example_knockoffs.py:139]:
```suggestion
# Knockoffs select the same variable as previously and do not select **any fake variable**.
```

File: examples/plot_example_knockoffs.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
+"""
+Controlled variable selection: why and how?
+=======================================
+"""
+# In this example, we explore the basics of variable selection and illustrate the need to statistically control the amount of faslely selected variables. We start off by loading the Wisconsin breast cancer dataset:
```
Comment (jpaillard) [inline examples/plot_example_knockoffs.py:5]: I suggest you use this code block to render the text nicely with sphinx. 
```python
"""
Controlled variable selection: why and how?
=======================================
In this example, we explore the basics of variable selection and illustrate the need to statistically control the amount of faslely selected variables. We start off by loading the Wisconsin breast cancer dataset:
"""
```

File: examples/plot_example_knockoffs.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+data = load_breast_cancer()
+X = data.data
+y = data.target
+
+X = StandardScaler().fit_transform(X)
+
+print(X.shape)
+n, p = X.shape
+
+
```
Comment (jpaillard) [inline examples/plot_example_knockoffs.py:27]: Add a new section to make the text appear in a text block and note a code comment. 
```suggestion

#############################################################################
```

**Review by jpaillard — COMMENTED** [review]
Thanks. I think it would be nice to have a plot in the example. It could illustrate the selected variables or illustrate that the FDR is controlled (or not) with knockoffs (or Lasso coef)

File: examples/plot_example_knockoffs.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+# Since we do not have the ground truth, we cannot evaluate this selection set directly. To investigate the reliability of this method, we artificially increase the number of variables by adding synthetic noise features. These are completely irrelevant to the problem at hand.
+
+#%%
+repeats_noise = 3  # Number of synthetic noisy sets to add
+
+X_to_shuff = X.copy()
+noises = [X]
+
+for k in range(repeats_noise):
+    np.random.shuffle(X_to_shuff.T)
```
Comment (jpaillard) [inline examples/plot_example_knockoffs.py:66]: also, here, I don't get why X is shuffled. It is not appended to `noises`

File: examples/plot_example_knockoffs.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+print("Selected features with Knockoffs:")
+print(np.array(feature_names)[selected])
+
+#%%
+# We now apply the Knockoffs procedure to the noisy dataset:
+
+#%%
+selected, test_scores, threshold, X_tildes = model_x_knockoff(
+    X,
+    y,
```
Comment (jpaillard) [inline examples/plot_example_knockoffs.py:119]: the final knockoff is actually done on the non-noisy data
